src/core_plugins/kibana/server/lib/crypto.js

```diff
@@ -0,0 +1,106 @@
+import crypto from 'crypto';
+
+const IV_LENGTH_IN_BYTES = 12;
+const SALT_LENGTH_IN_BYTES = 64;
+const KEY_LENGTH_IN_BYTES = 32;
+const KEY_ITERATIONS = 10000;
+const KEY_DIGEST = 'sha512';
+const CIPHER_ALGORITHM = 'aes-256-gcm';
+const ENCRYPTION_RESULT_ENCODING = 'base64';
+
+function _validateOpts(opts) {
+  if (!opts.encryptionKey) {
+    throw new Error('encryptionKey is required');
+  }
+}
+
+function _generateSalt() {
+  return crypto.randomBytes(SALT_LENGTH_IN_BYTES);
+}
+
+function _generateIV() {
+  return crypto.randomBytes(IV_LENGTH_IN_BYTES);
+}
+
+function _generateKey(encryptionKey, salt) {
+  if (!Buffer.isBuffer(salt)) {
+    salt = new Buffer(salt, ENCRYPTION_RESULT_ENCODING);
+  }
+
+  return new Promise((resolve, reject) => {
+    crypto.pbkdf2(encryptionKey, salt, KEY_ITERATIONS, KEY_LENGTH_IN_BYTES, KEY_DIGEST, (err, key) => {
+      if (err) {
+        reject(err);
+        return;
+      }
+
+      if (!Buffer.isBuffer(key)) {
+        key = new Buffer(key, 'binary');
+      }
+
+      resolve(key);
+    });
+  });
+}
+
+function _serialize(obj) {
+  return new Promise((resolve, reject) => {
+    const serializedObj = JSON.stringify(obj);
+    if (serializedObj === undefined) {
+      reject(new Error('Object to be encrypted must be serializable'));
+      return;
+    }
+    resolve(serializedObj);
+  });
+}
+
+/**
+ * Implmenetation of encrypt() and decrypt() taken from https://gist.github.com/AndiDittrich/4629e7db04819244e843,
+ * which was recommended by @jaymode
+ */
+export default function makeCryptoWith(opts) {
+
+  _validateOpts(opts);
+  const encryptionKey = opts.encryptionKey;
```

Comment:
My preference is to pass required parameters as arguments, and optional parameters as an object. The encryption key will *always* be required, so wrapping it in an object introduces additional required domain knowledge as you now need to know the parameter to include in the object.

Just my 2 cents.